src/rgw/rgw_gc.cc

```diff
@@ -65,7 +65,64 @@ int RGWGC::tag_index(const string& tag)
   return rgw_shards_mod(XXH64(tag.c_str(), tag.size(), seed), max_objs);
 }
 
-int RGWGC::send_chain(cls_rgw_obj_chain& chain, const string& tag)
+int RGWGC::send_split_chain(const cls_rgw_obj_chain& chain, const std::string& tag)
+{
+  constexpr int FACTOR_INCREASE = 3; //approximate value by which total size of chain, tag and time increases after encoding into bufferlist
+  ldpp_dout(this, 20) << "RGWGC::send_split_chain - tag is: " << tag << dendl;
+
+  if (cct->_conf->osd_max_write_size) {
+    cls_rgw_obj_chain broken_chain;
+    bool chain_sent = false;
+    long unsigned int tag_size = tag.length(), expiration_time_size = sizeof(ceph::real_time);
+    long unsigned int adjusted_osd_max_write_size = ((cct->_conf->osd_max_write_size << 20)/FACTOR_INCREASE); //approx upper limit
+    ldpp_dout(this, 20) << "RGWGC::send_split_chain - adjusted_osd_max_write_size is: " << adjusted_osd_max_write_size << dendl;
+    long unsigned int cumulative_size = 0;
+    for (auto it = chain.objs.begin(); it != chain.objs.end(); it++) {
+      long unsigned int size = it->pool.size() + it->loc.size() + it->key.name.size() + it->key.instance.size();
+      cumulative_size += size;
+      long unsigned int total_size = cumulative_size + tag_size + expiration_time_size;
+      ldpp_dout(this, 20) << "RGWGC::send_split_chain - cumulative_size is: " << cumulative_size << dendl;
+      ldpp_dout(this, 20) << "RGWGC::send_split_chain - total_size is: " << total_size << dendl;
+
+      if (total_size < adjusted_osd_max_write_size) { //continue adding to chain
+        broken_chain.objs.emplace_back(*it);
+        chain_sent = false;
+        ldpp_dout(this, 20) << "RGWGC::send_split_chain - less than, add to broken chain and continue" << dendl;
+        continue;
+      }
+      if (total_size == adjusted_osd_max_write_size) { //add to chain and send to gc
+        broken_chain.objs.emplace_back(*it);
+        ldpp_dout(this, 20) << "RGWGC::send_split_chain - equal to, add to broken chain and send chain" << dendl;
+        auto ret =  send_split_chain(broken_chain, tag);
+        if (ret < 0) {
+          ldpp_dout(this, 0) << "RGWGC::send_split_chain - send chain returned error: " << ret << dendl;
+          return ret;
+        }
+        chain_sent = true;
+      }
+      if (total_size > adjusted_osd_max_write_size) { //dont add to chain, and send to gc
+        --it;
+        ldpp_dout(this, 20) << "RGWGC::send_split_chain - more than, dont add to broken chain and send chain" << dendl;
+        auto ret = send_chain(broken_chain, tag);
+        if (ret < 0) {
```

Comment:
Before the change introduced by https://github.com/ceph/ceph/pull/28421, this strategy may work since gc list is in omap, which has no size constraint. 
But when we have https://github.com/ceph/ceph/pull/28421, rgw_gc_max_queue_size(osd_max_object_size - rgw_gc_max_deferred_entries_size - 1K) seems another factor to make the operation fail. And if the operation fails, the caller may fallback to delete the tail in a sync way. However, if the original S3 object is as large as several TBs, it may take minutes and even hours to delete its tail, which will result in HTTP timeout of RGWDeleteObj or RGWAbortMultipart

Reply:
@cbodley @pritha-srivastava @mattbenjamin 
Why don't we change the format of cls_rgw_obj_chain like this?
The advantage here is rgw will not care about osd_max_object_size.
The major disadvantage here, I think is we are assuming the pool mapping in the placement rule will not be change or deleted before GC works on the tail of the deleted object. Is it the major reason that we choose not to put Manifest in cls_rgw_obj_chain?

```
diff --git a/src/cls/rgw/cls_rgw_types.h b/src/cls/rgw/cls_rgw_types.h
index 4edaa73b303..a2271cc8fa8 100644
--- a/src/cls/rgw/cls_rgw_types.h
+++ b/src/cls/rgw/cls_rgw_types.h
@@ -1017,7 +1017,19 @@ WRITE_CLASS_ENCODER(cls_rgw_obj)
 struct cls_rgw_obj_chain {
   list<cls_rgw_obj> objs;

-  cls_rgw_obj_chain() {}
+  // when refer_to_m==true, objs is empty and
+  // we need to get calculate the  list from m_bl
+  bool refer_to_m;
+
+  //m_bl is a bufferlist to hold a rgw_raw_obj and a RGWObjManifest
+  // Use bufferlist instead of rgw_raw_obj and RGWObjManifest here
+  // because we want to minimize header file dependencies of cls implementations.
+  // Use m_bl instead of `objs` because `objs` will be hundreds of MBs
+  // (even as large as several GBs) when the the size of a s3 object is to TB level
+  // and osd_max_write_size will stop large `objs` from being persisted.
+  bufferlist m_bl;
+
+  cls_rgw_obj_chain(): refer_to_m(false) {}

   void push_obj(const string& pool, const cls_rgw_obj_key& key, const string& loc) {
     cls_rgw_obj obj;
@@ -1025,17 +1037,24 @@ struct cls_rgw_obj_chain {
     obj.key = key;
     obj.loc = loc;
     objs.push_back(obj);
+    refer_to_m = false;
   }

   void encode(bufferlist& bl) const {
-    ENCODE_START(1, 1, bl);
+    ENCODE_START(2, 1, bl);
     encode(objs, bl);
+    encode(refer_to_m, bl);
+    encode(m_bl, bl);
     ENCODE_FINISH(bl);
   }

   void decode(bufferlist::const_iterator& bl) {
-    DECODE_START(1, bl);
+    DECODE_START(2, bl);
     decode(objs, bl);
+    if (struct_v >= 2) {
+      decode(refer_to_m, bl);
+      decode(m_bl, bl);
+    }
     DECODE_FINISH(bl);
   }

@@ -1053,7 +1072,14 @@ struct cls_rgw_obj_chain {
   }

   bool empty() {
-    return objs.empty();
+    return objs.empty() && !refer_to_m;
+  }
+
+  void set_manifest(const bufferlist& obj_manifest_bl)
+  {
+    m_bl = obj_manifest_bl;
+    refer_to_m = true;
+    objs.clear();
   }
 };
 WRITE_CLASS_ENCODER(cls_rgw_obj_chain)

```

Reply:
thanks @Jeegn-Chen, i do think it's worth experimenting with that change in a separate PR. that encoded-manifest strategy only works once every RGW and OSD upgrades, but this PR works over rolling upgrades of RGWs so it's better in the short term

even with support for manifest-based gc chains, it's likely that some objects (small or multipart?) would use less space in their existing `list<cls_rgw_obj>` form. we should put some thought into a heuristic to decide between the two